src/components/encoding-pane/property-editor-schema.ts

```diff
@@ -0,0 +1,328 @@
+  'ui:widget': 'select',
+  'ui:placeholder': 'auto',
+  'ui:emptyValue': 'auto'
+};
+
+// ------------------------------------------------------------------------------
+// Channel-Field Indexes for custom encoding
+// Key is Tab name, value is list of fieldDef properties
+// ------------------------------------------------------------------------------
+
+const AXIS_ORIENT_TITLE = ['title', 'orient'].map(p => ({prop: 'axis', nestedProp: p}));
+const LEGEND_ORIENT_TITLE = ['orient', 'title'].map(p => ({prop: 'legend', nestedProp: p}));
+
+const POSITION_FIELD_NOMINAL_INDEX = {
+  'Common': [
+    {
+      prop: 'scale',
+      nestedProp: 'type'
+    },
+    ...AXIS_ORIENT_TITLE
+  ]
+};
+
+// TODO: this should differ in the future
+const POSITION_FIELD_TEMPORAL_INDEX = POSITION_FIELD_NOMINAL_INDEX;
+
+const POSITION_FIELD_QUANTITATIVE_INDEX = {
+  'Common': [
+    ...POSITION_FIELD_NOMINAL_INDEX['Common'],
+    {prop: 'stack'}
+  ]
+};
+
+const COLOR_CHANNEL_FIELD_INDEX = {
+  'Legend': ['orient', 'title', 'type'].map(p => ({prop: 'legend', nestedProp: p})),
+  'Scale': ['type', 'domain', 'scheme'].map(p => ({prop: 'scale', nestedProp: p}))
+};
+
+const SIZE_CHANNEL_FIELD_INDEX = {
+  'Legend': ['orient', 'title'].map(p => ({prop: 'legend', nestedProp: p})),
+  'Scale': ['type', 'domain', 'range'].map(p => ({prop: 'scale', nestedProp: p}))
+};
+
+const SHAPE_CHANNEL_FIELD_INDEX = {
+  'Legend': LEGEND_ORIENT_TITLE,
+  // No need to adjust scale type for shape
+  'Scale': ['domain', 'range'].map(p => ({prop: 'scale', nestedProp: p}))
+};
+
+// ------------------------------------------------------------------------------
+// Color Scheme Constants
+// ------------------------------------------------------------------------------
+export const CATEGORICAL_COLOR_SCHEMES = ['accent', 'category10', 'category20', 'category20b', 'category20c', 'dark2',
+  'paired', 'pastel1', 'pastel1', 'set1', 'set2', 'set3', 'tableau10', 'tableau20'];
+
+export const SEQUENTIAL_COLOR_SCHEMES = ['blues', 'greens', 'greys', 'purples', 'reds', 'oranges', 'viridis', 'inferno',
+  'magma', 'plasma', 'bluegreen', 'bluepurple', 'greenblue', 'orangered', 'blueorange'];
+
+// ------------------------------------------------------------------------------
+// Generator/Factory Methods
+// ------------------------------------------------------------------------------
+export function generatePropertyEditorSchema(prop: string, nestedProp: string, propTab: string,
+                                             fieldDef: ShelfFieldDef, shelfId: ShelfId): PropertyEditorSchema {
+  const title = generateTitle(prop, nestedProp, propTab);
+  const baseUISchema: SchemaProperty = {
+    title: title,
+    type: 'string'
+  };
+
+  const {channel} = shelfId;
+
+  if (prop === 'scale') {
+    if (nestedProp === 'type') {
+      // Filtering based on channel type & fieldDef type
+      const scaleTypes: string[] = getSupportedScaleTypes(shelfId, fieldDef);
+      const propertySchema = {
```

Comment:
@ssharif6 The need to declare this propertySchema and call generateSchema with so many parameters is every time makes this method a bit painful to read. 

We should 
1) break this method to multiple method (`generateScale/Axis/LegendEditorSchema`)

2) use `switch-case` instead of `if`  since these are all exact match comparison

3) just pass `propertyKey` -- we don't really need `prop: string, nestedProp: string` inside?

4) Make specialized forms of generateSchema. 

- For example, enum is always used with select box.  So you should instead have 
`generateSelectSchema(propKey, title, enum)`. 

- Similarly, `generateTextBoxSchema(propKey, title, 'string' | 'number')`. 


So there are way less parameters and easier to read / maintain.